tests/Locator/RootsFileOrDirectoryLocatorTest.php

```diff
@@ -0,0 +1,437 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Locator;
+
+use Generator;
+use Infection\Locator\FileNotFound;
+use Infection\Locator\RootsFileOrDirectoryLocator;
+use function Infection\Tests\normalizePath as p;
+use function iterator_to_array;
+use PHPUnit\Framework\TestCase;
+use function Safe\realpath;
+use function Safe\sprintf;
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * @internal
+ * @group integration
+ */
+final class RootsFileOrDirectoryLocatorTest extends TestCase
+{
+    private const FIXTURES_DIR = __DIR__ . '/../Fixtures/Locator';
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp(): void
+    {
+        $this->filesystem = new Filesystem();
+    }
+
+    /**
+     * @dataProvider pathsProvider
+     */
+    public function test_it_can_locate_files(array $roots, string $file, string $expected): void
+    {
+        $path = (new RootsFileOrDirectoryLocator($roots, $this->filesystem))->locate($file);
+
+        $this->assertSame(p($expected), p($path));
+    }
+
+    /**
+     * @dataProvider invalidPathsProvider
+     */
+    public function test_it_throws_an_exception_if_file_or_folder_does_not_exist(
+        array $roots,
+        string $file,
+        string $expectedErrorMessage
+    ): void {
+        $locator = new \Infection\Locator\RootsFileOrDirectoryLocator($roots, $this->filesystem);
+
+        try {
+            $locator->locate($file);
+
+            $this->fail();
```

Comment:
What do you think about moving this statement out of the `try`-block, and instead using a `return` statement in the `catch` block after successful inspection of the expected exception?

Reply:
I personally find it clearer like that IMO. The flow is harder to follow if you rely on a return statement here and as error pron

Reply:
For the reader it’s not clear though which of these statements might produce an exception.

Reply:
I don't think it's causing any confusion because you're not supposed to be able to catch PHPUnit's exceptions that way, neither people are necessary aware that `fail()` actually just throw a specific exception. To me it reads more clearly as "I've reached that line but that shouldn't have happened".

That being said I called `$this->fail()` out of habit but I guess I can try to be more friendly here like I did in other tests and put a more helpful error message